base/abstractarray.jl

```diff
@@ -373,9 +373,16 @@ zero{T}(x::AbstractArray{T}) = fill!(similar(x), zero(T))
 # While the definitions for LinearFast are all simple enough to inline on their
 # own, LinearSlow's CartesianRange is more complicated and requires explicit
 # inlining.
-start(A::AbstractArray) = (@_inline_meta(); itr = eachindex(A); (itr, start(itr)))
-next(A::AbstractArray,i) = (@_inline_meta(); (idx, s) = next(i[1], i[2]); (A[idx], (i[1], s)))
-done(A::AbstractArray,i) = done(i[1], i[2])
+start(A::AbstractArray) = (@_inline_meta(); _start(linearindexing(A), A))
+_start(::LinearFast, A::AbstractArray) = 1
+_start(::LinearSlow, A::AbstractArray) = (@_inline_meta(); itr = eachindex(A); (itr, start(itr)))
```

Comment:
It doesn't matter what type it was.  If you get a `pos` from `start(A)`, then I firmly believe the only things you can do with it are call `next` and `done`.  If you want to do linear arithmetic with `pos`, then I think you should start with `pos = 1`.  If you want to support non-1-based arrays, then maybe we can discuss a new method `pos = first_linear_index(A)`.